chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/ContourDataSetRenderer.java

```diff
@@ -229,82 +170,105 @@ private void drawContourFast(final GraphicsContext gc, final AxisTransform axisT
-            // erosionOperator2(output2, output, zMin, zMax, levels[i]);
-
-            for (int xIndex = indexXMin; xIndex < indexXMax; xIndex++) {
-                for (int yIndex = indexYMin; yIndex < indexYMax; yIndex++) {
-                    // final double z = dataSet3D.getZ(xIndex, yIndex);
-                    final double z = output[xIndex - indexXMin][yIndex - indexYMin];
-
-                    // Color color = getColor(z);
-                    Color color = lCache.zInverted ? getColor(1 - z) : getColor(z);
-                    if (z > 0) {
-                        color = lCache.zInverted ? getColor(1 - level) : getColor(level);
-                    } else {
-                        color = Color.TRANSPARENT;
+
+            for (int yIndex = 0; yIndex < ySize; yIndex++) {
+                final int yIndex2 = ySize - 1 - yIndex;
+                for (int xIndex = 0; xIndex < xSize; xIndex++) {
+                    final double z = output[xIndex][yIndex];
+
+                    if (z <= 0) {
                         continue;
                     }
+                    Color color = lCache.zInverted ? colorGradient.getColor(1 - level) : colorGradient.getColor(level);
 
-                    for (int dx = 0; dx < scaleX; dx++) {
-                        for (int dy = 0; dy < scaleY; dy++) {
-                            final int x = (xIndex - indexXMin) * scaleX;
-                            final int y = (indexYMax - 1 - yIndex) * scaleY;
-                            pixelWriter.setColor(x + dx, y + dy, color);
-                        }
-                    }
+                    pixelWriter.setColor(xIndex, yIndex2, color);
                 }
             }
         }
 
-        gc.drawImage(image, 0, 0, lCache.xAxisWidth, lCache.yAxisHeight);
+        gc.drawImage(image, lCache.xDataPixelMin, lCache.yDataPixelMin, lCache.xDataPixelRange, lCache.yDataPixelRange);
         ProcessingProfiler.getTimeDiff(start, "sobel");
     }
 
-    private void drawHeatMap(final GraphicsContext gc, final AxisTransform axisTransform, final Cache lCache) {
+    private void drawHeatMap(final GraphicsContext gc, final ContourDataSetCache lCache) {
+        final long start = ProcessingProfiler.getTimeStamp();
+
+        // N.B. works only since OpenJFX 12!! fall-back for JDK8 is the old implementation
+        gc.setImageSmoothing(isSmooth());
+
+        // process z quantisation to colour transform
+        final Image image = ContourDataSetCache.convertDataArrayToImage(lCache.reduced, lCache.xSize, lCache.ySize,
+                getColorGradient());
+        ProcessingProfiler.getTimeDiff(start, "color map");
+
+        gc.drawImage(image, lCache.xDataPixelMin, lCache.yDataPixelMin, lCache.xDataPixelRange, lCache.yDataPixelRange);
+
+        ProcessingProfiler.getTimeDiff(start, "drawHeatMap");
+    }
+
+    private void drawHeatMapOld(final GraphicsContext gc, final AxisTransform axisTransform,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'drawHeatMapOld(GraphicsContext, AxisTransform, ContourDataSetCache)' has an NPath complexity of 288](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=4880492)